### PR TITLE
bugfix: In case of err in vmrestore, leave VM without RestoreInProgress annotation allowing it to be started

### PR DESCRIPTION
**Pull request by ShellyKa13**
### What this PR does
Before this PR: VMRestore gets deleted without any cleanup of validation it can be deleted.

After this PR:     When calling delete on the restore we must make sure we are not leaving on the vm the restoreInProgress annotation. For that added a finalizer on the vmrestore object making sure it is not deleted until we remove the annotation from the vm. Queued the vmrestore object in case of deletion, then if the vm exists we remove the restore in progress annotation if it is for the same restore, only then remove restore finalizer allowing it to be deleted.

Fixes # Jira-ticket: https://issues.redhat.com/browse/CNV-47106

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
bugfix: In case of err in vmrestore, leave VM without RestoreInProgress annotation allowing it to be started
```

### CODE REVIEW

**Comment by ShellyKa13** [comment]
/hold
The flake is still happening and another test seem to fail. looking into that. Review is still welcomed

File: pkg/storage/snapshot/restore.go [hunk 1]
```diff
@@ -122,11 +125,19 @@ func (ctrl *VMRestoreController) updateVMRestore(vmRestoreIn *snapshotv1.Virtual
...
+	if !VmRestoreProgressing(vmRestoreOut) {
+		return 0, nil
+	}
+
 	if len(vmRestoreOut.OwnerReferences) == 0 {
 		target.Own(vmRestoreOut)
 		updateRestoreCondition(vmRestoreOut, newProgressingCondition(corev1.ConditionTrue, "Initializing VirtualMachineRestore"))
 		updateRestoreCondition(vmRestoreOut, newReadyCondition(corev1.ConditionFalse, "Initializing VirtualMachineRestore"))
 	}
+	controller.AddFinalizer(vmRestoreOut, vmRestoreFinalizer)
```
Comment (mhenriks) [inline pkg/storage/snapshot/restore.go:140]: Would be best to update the restore and return when the finalizer is added before even touching the VM. Could be a problem if `UpdateRestoreInProgress` succeeds but the restore update fails

File: pkg/storage/snapshot/restore.go [hunk 2]
```diff
@@ -236,6 +247,29 @@ func (ctrl *VMRestoreController) doUpdate(original, updated *snapshotv1.VirtualM
 	return nil
 }
 
+func (ctrl *VMRestoreController) handleVMRestoreDeletion(vmRestore *snapshotv1.VirtualMachineRestore, target restoreTarget) error {
+	logger := log.Log.Object(vmRestore)
```
Comment (mhenriks) [inline pkg/storage/snapshot/restore.go:251]: if resource does not have the finalizer should the function exit?

File: pkg/storage/snapshot/restore.go [hunk 3]
```diff
@@ -363,6 +401,8 @@ func (t *vmRestoreTarget) UpdateRestoreInProgress() error {
 
 		// unfortunately, status Updater does not return the updated resource
 		// but the controller is watching VMs so will get notified
+		// TODO: if this is true then don't we need to exist the reconcile loop
+		// to make sure we are working on an updated vm?
```
Comment (mhenriks) [inline pkg/storage/snapshot/restore.go:405]: I think it only matters if the VM will be updated again in the same reconcile loop, that should not be the case

**Review by mhenriks — COMMENTED** [review]
initial comments

File: pkg/storage/snapshot/restore.go (hunk 3 above)
Comment (mhenriks) [reply pkg/storage/snapshot/restore.go:405]: actually now that no longer using the "status updater" abstraction can get rid of comment and get updated vm returned

**Comment by ShellyKa13** [comment]
/unhold
Quick fix for the flake did not work. Removed the commit with this change and the unquarantine of the quarantined tests. Will look deeper into that and open a new PR with the fix to prevent keeping the bugfix from hanging longer.
@mhenriks fixed your comments, if you can have another look, thanks!

File: tests/storage/restore.go [hunk 4]
```diff
@@ -1379,10 +1383,17 @@ var _ = SIGDescribe("VirtualMachineRestore Tests", func() {
...
+				case "deleteWebhook":
+					deleteWebhook(webhook)
+					webhook = nil
 
-				restore = waitRestoreComplete(restore, vm.Name, &vm.UID)
+					restore = waitRestoreComplete(restore, vm.Name, &vm.UID)
+				case "deleteRestore":
+					deleteRestore(restore)
+				default:
+					panic("Delete function not valid")
```
Comment (akalenyu) [inline tests/storage/restore.go:1395]: Maybe `Fail()` is cleaner here? so ginkgo can perform some of it's logic for failures

File: pkg/storage/snapshot/restore.go [hunk 5]
```diff
@@ -174,7 +173,7 @@ func (ctrl *VMRestoreController) updateVMRestore(vmRestoreIn *snapshotv1.Virtual
 		return 0, ctrl.doUpdate(vmRestoreIn, vmRestoreOut)
 	}
 
-	if err = target.Cleanup(); err != nil {
+	if err = ctrl.deleteObsuleteDataVolumes(vmRestoreOut); err != nil {
```
Comment (akalenyu) [inline pkg/storage/snapshot/restore.go:187]: typo Obsulete->Obsolete

File: pkg/storage/snapshot/restore.go [hunk 6]
```diff
@@ -49,6 +49,8 @@ import (
 const (
 	RestoreNameAnnotation = "restore.kubevirt.io/name"
 
+	vmRestoreFinalizer = "snapshot.kubevirt.io/vmrestore-protection"
```
Comment (akalenyu) [inline pkg/storage/snapshot/restore.go:52]: Just a note about finalizers, we used to have a bug where we forget to remove one, so the resource could be left dangling. I don't see how it's possible with this PR but just for awareness https://github.com/kubevirt/kubevirt/pull/11241/commits/6ad91329f2e72aaa6b6c961fdf38496052602621

**Review by akalenyu — COMMENTED** [review]
looks good

File: pkg/storage/snapshot/restore.go (hunk 6 above)
Comment (mhenriks) [reply pkg/storage/snapshot/restore.go:52]: Should make sure finalizers are removed here as well, for example: https://github.com/kubevirt/kubevirt/blob/main/hack/cluster-clean.sh#L49-L54

**Comment by ShellyKa13** [comment]
Ran the UT 30 times locally with cleaning cache and it worked. Making sure it doesnt flake in CI either:
/test pull-kubevirt-unit-test

**Comment by akalenyu** [comment]
/test pull-kubevirt-e2e-k8s-1.31-sig-storage
working on etcd errors async with @brianmcarey https://github.com/kubevirt/project-infra/pull/3626#issuecomment-2358664646

**Comment by akalenyu** [comment]
```
07:50:04:   [PANICKED] Test Panicked
07:50:04:   In [SynchronizedAfterSuite] at: tests/util/util.go:5 @ 09/19/24 07:50:04.448
07:50:04: 
07:50:04:   Internal error occurred: etcdserver: request timed out
```
@brianmcarey can we please override?

**Comment by brianmcarey** [comment]
> ```
> 07:50:04:   [PANICKED] Test Panicked
> 07:50:04:   In [SynchronizedAfterSuite] at: tests/util/util.go:5 @ 09/19/24 07:50:04.448
> 07:50:04: 
> 07:50:04:   Internal error occurred: etcdserver: request timed out
> ```
> 
> @brianmcarey can we please override? This passes on the other two lanes and failed on the aftersuite due to an etcd timeout infra flake

Ok thanks @akalenyu 

/override pull-kubevirt-e2e-k8s-1.30-sig-storage

**Pull request merged**